File: frontend/src/lib/dinero.ts
import Decimal from "decimal.js";

/**
 * Dinero en el frontend.
 *
 * La API manda los montos como string (`"359.67"`) y acá se parsean con decimal.js,
 * nunca con `Number`. No es purismo: en JavaScript `0.1 + 0.2 !== 0.3`, así que sumar
 * montos como floats produce exactamente los descuadres de un centavo que el panel de
 * conciliación tendría que estar detectando. Un total mal calculado en la pantalla es
 * peor que ninguno, porque se le cree.
 */

export type Monto = string;

export function d(valor: Monto | number | Decimal | null | undefined): Decimal {
  if (valor === null || valor === undefined || valor === "") return new Decimal(0);
  return new Decimal(valor as Decimal.Value);
}

/** Suma una lista de montos sin pasar por float. */
export function sumar(valores: Array<Monto | null | undefined>): Decimal {
  return valores.reduce((acc, v) => acc.plus(d(v)), new Decimal(0));
}

const formatoUsd = new Intl.NumberFormat("en-US", {
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
});

const formatoBs = new Intl.NumberFormat("es-VE", {
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
});

/**
 * `$359.67`. Convención estadounidense: coma de miles, punto decimal.
 *
 * Las dos convenciones conviven en este negocio y mezclarlas es el bug que encontró
 * la auditoría del Excel, así que cada moneda tiene su formateador.
 */
export function usd(valor: Monto | number | Decimal | null | undefined): string {
  return `$${formatoUsd.format(d(valor).toNumber())}`;
}

/** `Bs. 15.679,22`. Convención venezolana: punto de miles, coma decimal. */
export function bs(valor: Monto | number | Decimal | null | undefined): string {
  return `Bs. ${formatoBs.format(d(valor).toNumber())}`;
}

/** El número sin símbolo, para inputs. */
export function plano(valor: Monto | number | Decimal | null | undefined): string {
  return d(valor).toFixed(2);
}

export function porcentaje(
  valor: Monto | number | null | undefined,
  decimales = 1,
): string {
  return `${d(valor).times(100).toFixed(decimales)}%`;
}

/**
 * Reparte un total en cuotas exactas: el resto va a la última.
 *
 * Es la misma regla que el backend (`repartir` en core/dinero.py) para que el preview
 * del plan y lo que se guarda no puedan discrepar.
 */
export function repartir(total: Monto | Decimal, partes: number): Decimal[] {
  if (partes < 1) throw new Error("Hacen falta al menos 1 parte.");
  const t = d(total);
  const base = t.div(partes).toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
  const cuotas = Array.from({ length: partes - 1 }, () => base);
  cuotas.push(t.minus(base.times(partes - 1)).toDecimalPlaces(2, Decimal.ROUND_HALF_UP));
  return cuotas;
}

export { Decimal };

File: backend/app/core/dinero.py
"""Aritmética de dinero. Decimal siempre, float nunca.

`ROUND_HALF_UP` y no el banquero de Python por defecto: es lo que hace el negocio
a mano y lo que hacía el Excel, y una diferencia de medio centavo repetida es
exactamente lo que el panel de conciliación tendría que estar detectando.
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal

CENTAVO = Decimal("0.01")
EPSILON = Decimal("0.005")
UNO = Decimal("1")


def a_decimal(valor: object) -> Decimal:
    if isinstance(valor, Decimal):
        return valor
    if isinstance(valor, float):
        return Decimal(str(valor))
    return Decimal(str(valor or "0"))


def cuantizar(valor: object) -> Decimal:
    """Redondea a 2 decimales, medio hacia arriba."""
    return a_decimal(valor).quantize(CENTAVO, rounding=ROUND_HALF_UP)


def convertir_a_usd(monto_moneda: object, tasa: object) -> Decimal:
    """monto / tasa, cuantizado. La tasa tiene que ser positiva."""
    tasa_dec = a_decimal(tasa)
    if tasa_dec <= 0:
        raise ValueError("La tasa tiene que ser mayor que cero.")
    return cuantizar(a_decimal(monto_moneda) / tasa_dec)


def convertir_a_moneda(monto_usd: object, tasa: object) -> Decimal:
    tasa_dec = a_decimal(tasa)
    if tasa_dec <= 0:
        raise ValueError("La tasa tiene que ser mayor que cero.")
    return cuantizar(a_decimal(monto_usd) * tasa_dec)


def es_cero(valor: object) -> bool:
    """Cero con la tolerancia de medio centavo que ya usaba register_payment."""
    return abs(a_decimal(valor)) < EPSILON


def repartir(total: object, partes: int) -> list[Decimal]:
    """Divide en `partes` cuotas exactas: el resto va a la última.

    repartir('90.00', 3)  -> [30.00, 30.00, 30.00]
    repartir('100.00', 3) -> [33.33, 33.33, 33.34]
    """
    if partes < 1:
        raise ValueError("Hacen falta al menos 1 parte.")
    total_dec = cuantizar(total)
    base = cuantizar(total_dec / partes)
    cuotas = [base] * (partes - 1)
    cuotas.append(cuantizar(total_dec - base * (partes - 1)))
    return cuotas


def formatear_usd(valor: object) -> str:
    return f"${cuantizar(valor):,.2f}"


def formatear_bs(valor: object) -> str:
    """Convención venezolana: punto de miles, coma decimal."""
    crudo = f"{cuantizar(valor):,.2f}"
    return "Bs. " + crudo.replace(",", "\x00").replace(".", ",").replace("\x00", ".")
